streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java

```diff
@@ -125,35 +125,41 @@ public boolean isValidTransition(final State newState) {
     boolean commitNeeded();
 
     /**
-     * @throws TaskMigratedException all the task has been migrated
      * @throws StreamsException fatal error, should close the thread
      */
-    void commit();
+    void prepareCommit();
+
+    void postCommit();
 
     /**
      * @throws TaskMigratedException all the task has been migrated
      * @throws StreamsException fatal error, should close the thread
      */
-    void suspend();
+    void prepareSuspend();
 
+    void suspend();
     /**
+     *
      * @throws StreamsException fatal error, should close the thread
      */
     void resume();
 
     /**
-     * Close a task that we still own. Commit all progress and close the task gracefully.
+     * Prepare to close a task that we still own and prepare it for committing
```

Comment:
nit: we should emphasize that PrepareClose and close calls should be implemented idempotent since we may call it multiple times if a task close clean first and then fail and then close dirty.